src/source/fadeout.rs

```diff
@@ -0,0 +1,95 @@
+use std::time::Duration;
+
+use Sample;
+use Source;
+
+/// Internal function that builds a `FadeOut` object.
+pub fn fadeout<I>(input: I, duration: Duration) -> FadeOut<I> {
+    let duration = duration.as_secs() * 1000000000 + duration.subsec_nanos() as u64;
+
+    FadeOut {
+        input: input,
+        remaining_ns: duration as f32,
+        total_ns: duration as f32,
+    }
+}
+
+/// Filter that modifies reduces the volume to silence over a time period.
+#[derive(Clone, Debug)]
+pub struct FadeOut<I> {
+    input: I,
+    remaining_ns: f32,
+    total_ns: f32,
+}
+
+impl<I> FadeOut<I> {
+    /// Starts the fade to silence.
+    #[inline]
+    pub fn start(&mut self) {
+        self.remaining_ns = self.total_ns;
+    }
+
+    /// Clears the fade out time.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.remaining_ns = -1.0;
+    }
+}
+
+impl<I> Iterator for FadeOut<I>
+where
+    I: Source,
+    I::Item: Sample,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<I::Item> {
+        if self.remaining_ns <= 0.0 {
+            return self.input.next();
```

Comment:
So once we reach 0.0, we will play full volume again?

Reply:
Ooh, good point.